frontend/src/presentation/components/QuestionList.tsx

```diff
@@ -1,83 +1,124 @@
-import React, { useState } from "react";
-import { List, Spin, Alert } from "antd";
-import { Question } from "../../domain/entities/Question";
-import { QuestionCard } from "./QuestionCard";
-import styles from "./QuestionList.module.css";
-import { SearchBar } from "./SearchBar";
+import React, { useState, useMemo, useCallback } from 'react';
+import { List, Spin, Alert } from 'antd';
+import { Question } from '../../domain/entities/Question';
+import { QuestionCard } from './QuestionCard';
+import { QuestionFilters } from './QuestionFilters';
+import styles from './QuestionList.module.css';
 
 interface QuestionListProps {
-	questions: Question[];
-	selectedQuestion: Question | null;
-	onSelectQuestion: (questionId: string) => void;
-	isNarrow: boolean;
-	isLoading: boolean;
-	error: string | null;
+  questions: Question[];
+  selectedQuestionId: string | null;
+  onSelectQuestion: (questionId: string) => void;
+  isNarrow: boolean;
+  isLoading: boolean;
+  error: string | null;
 }
 
 export const QuestionList: React.FC<QuestionListProps> = ({
-	questions,
-	selectedQuestion,
-	onSelectQuestion,
-	isNarrow,
-	isLoading,
-	error,
+  questions,
+  selectedQuestionId,
+  onSelectQuestion,
+  isNarrow,
+  isLoading,
+  error,
 }) => {
-	const [searchTerm, setSearchTerm] = useState("");
+  const [filters, setFilters] = useState({
+    selectedDifficulty: 'All',
+    selectedCategories: [] as string[],
+    searchTerm: '',
+  });
 
-	const filteredQuestions = questions.filter((question) => {
-		const matchesSearch =
-			question.title.toLowerCase().includes(searchTerm.toLowerCase()) ||
-			question.description.toLowerCase().includes(searchTerm.toLowerCase());
+  //TODO: Change this into something less hacky. It currently just 'XORs' all question categories to get
+  //a complete list. Ideally, there should just be a list of categories in the backend to query.
+  const allCategories = useMemo(() => {
+    const categoriesSet = new Set<string>();
+    questions.forEach((question) => {
+      question.categories.forEach((category) => categoriesSet.add(category));
+    });
+    return Array.from(categoriesSet);
+  }, [questions]);
 
-		return matchesSearch;
-	});
+  const handleFiltersChange = useCallback((newFilters: React.SetStateAction<{ 
+		selectedDifficulty: string; 
+		selectedCategories: string[]; 
+		searchTerm: string; }>) => {
+    setFilters(newFilters);
+  }, []);
 
-	const handleSearch = (term: string) => {
-		setSearchTerm(term);
-	};
+  const filteredQuestions = useMemo(() => {
+    return questions.filter((question) => {
+      if (
+        filters.selectedDifficulty !== 'All' &&
```

Comment:
Hmm, I feel checking `filters?.selectedDifficulty === 'All' && filters?.selectedCategories?.length === 0` then returning true at the top makes the logic inside a bit more clearer. What do u think?

Reply:
It would miss the search term filter. The other lines of code for filtering by selected difficulty and categories are still also still needed, making your suggestion, while nicer to read, kind of redundant.